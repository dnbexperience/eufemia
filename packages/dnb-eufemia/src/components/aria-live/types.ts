export type AriaLiveProps = {
  element?: React.ElementType

  /**
   * The variant of the announcement. Can be 'text' or 'content'.
   */
  variant?: 'text' | 'content'

  /**
   * The priority level of the announcement. Can be 'low', or 'high'.
   */
  priority?: 'low' | 'high'

  /**
   * Delay in milliseconds before the announcement is made. Defaults to 1000.
   * @default 1000
   */
  delay?: number

  /**
   * Whether the AriaLive component is on or off.
   * @default false
   */
  disabled?: boolean

  /**
   * The politeness level of the announcement. Can be 'off', 'polite', or 'assertive'.
   */
  politeness?: React.HTMLAttributes<HTMLElement>['aria-live']

  /**
   * Whether the entire region should be considered as a whole when communicating updates.
   */
  atomic?: React.HTMLAttributes<HTMLElement>['aria-atomic']

  /**
   * What types of changes should be presented to the user. Can be 'additions', 'removals', 'text', or 'all'.
   */
  relevant?: React.HTMLAttributes<HTMLElement>['aria-relevant']

  /**
   * Whether to show the children or not.
   * @default false
   */
  showAnnouncement?: boolean

  /**
   * The content to be announced.
   */
  children: React.ReactNode
}

export type AriaLiveAllProps = AriaLiveProps &
  React.HTMLAttributes<HTMLElement>
