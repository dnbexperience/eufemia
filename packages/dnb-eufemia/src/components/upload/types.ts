import React from 'react'
import type { SkeletonShow } from '../skeleton/Skeleton'
import type { LocaleProps, SpacingProps } from '../../shared/types'

export type UploadAcceptedFileTypes = string[]

export type UploadAcceptedFileTypesWithFileMaxSize =
  UploadAcceptedFileTypeObject[]

export type UploadAcceptedFileTypeObject = {
  fileType: string
  fileMaxSize?: number | false
}

export type UploadProps = {
  /**
   * unique id used with the useUpload hook to manage the files
   */
  id: string

  /**
   * list of accepted file types.
   */
  acceptedFileTypes:
    | UploadAcceptedFileTypes
    | UploadAcceptedFileTypesWithFileMaxSize

  /**
   * Skeleton should be applied when loading content
   * Default: null
   */
  skeleton?: SkeletonShow

  /**
   * Defines the amount of files the user can select and upload
   * Default: 100
   */
  filesAmountLimit?: number

  /**
   * Defines the max file size of each file in MB. Use either `0` or `false` to disable.
   * Default: 5 MB
   */
  fileMaxSize?: number | false

  /**
   * will be called on `files` changes made by the user. Access the files with `{ files }`.
   */
  onChange?: ({ files }: { files: UploadFile[] }) => void

  /**
   * will be called once a file gets deleted by the user. Access the deleted file with `{ fileItem }`.
   */
  onFileDelete?: ({ fileItem }: { fileItem: UploadFile }) => void

  /**
   * Custom text properties
   */
  title?: React.ReactNode
  text?: React.ReactNode
  fileTypeTableCaption?: React.ReactNode
  fileTypeDescription?: React.ReactNode
  fileSizeDescription?: React.ReactNode
  fileAmountDescription?: React.ReactNode
  fileSizeContent?: React.ReactNode
  buttonText?: React.ReactNode
  errorLargeFile?: React.ReactNode
  errorUnsupportedFile?: React.ReactNode
  errorAmountLimit?: React.ReactNode
  loadingText?: React.ReactNode
  deleteButton?: React.ReactNode
  fileListAriaLabel?: string
  children?: React.ReactNode
}

export type UploadAllProps = UploadProps &
  SpacingProps &
  LocaleProps &
  Omit<React.HTMLProps<HTMLElement>, 'onChange' | 'title'>

export type UploadContextProps = {
  id?: string
  onInputUpload: (files: UploadFile[]) => void
} & Partial<UploadProps>

export type UploadFile = {
  file: File
  errorMessage?: React.ReactNode
  isLoading?: boolean
  exists?: boolean
  id?: string
}
