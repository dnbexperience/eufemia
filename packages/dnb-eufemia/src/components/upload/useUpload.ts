import { useEventEmitter } from '../../../src/shared/component-helper'
import type { UploadFile } from './types'

/**
 * Use together with Upload with the same id to manage the files from outside the component.
 * @param id string, must match the id of the Upload component
 * @returns { files: UploadFile[], setFiles: (file: UploadFile[]) => void }
 */
function useUpload(id: string): {
  files: UploadFile[]
  setFiles: (files: UploadFile[]) => void
} {
  const { data, update } = useEventEmitter(id)

  function setFiles(files: UploadFile[]) {
    update({ files })
  }

  return {
    files: data?.files || [],
    setFiles,
  }
}

export default useUpload
