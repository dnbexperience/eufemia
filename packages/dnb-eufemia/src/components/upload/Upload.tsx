import React from 'react'
import classnames from 'classnames'

// Shared
import { createSpacingClasses } from '../space/SpacingHelper'
import Provider from '../../shared/Provider'
import Context from '../../shared/Context'
import {
  extendPropsWithContext,
  makeUniqueId,
} from '../../shared/component-helper'

// Internal
import UploadFileInput from './UploadFileInput'
import useUpload from './useUpload'
import UploadDropzone from './UploadDropzone'
import { UploadContext, defaultProps } from './UploadContext'
import { verifyFiles } from './UploadVerify'

import type { UploadFile, UploadAllProps } from './types'
import UploadFileList from './UploadFileList'
import UploadInfo from './UploadInfo'

export { defaultProps }

const Upload = (localProps: UploadAllProps) => {
  const context = React.useContext(Context)

  const { buttonTextSingular, textSingular, ...translations } =
    context.getTranslation(localProps).Upload

  if (localProps?.filesAmountLimit === 1) {
    translations.buttonText = buttonTextSingular
    translations.text = textSingular
  }

  const extendedProps = extendPropsWithContext(
    localProps,
    defaultProps,
    { skeleton: context?.skeleton },
    translations,
    context.Upload
  )

  const {
    id,
    skeleton,
    className,
    acceptedFileTypes,
    filesAmountLimit,
    fileMaxSize,
    onChange,
    onFileDelete, // eslint-disable-line
    title, // eslint-disable-line
    text, // eslint-disable-line
    fileTypeDescription, // eslint-disable-line
    fileSizeDescription, // eslint-disable-line
    fileAmountDescription, // eslint-disable-line
    fileSizeContent, // eslint-disable-line
    buttonText, // eslint-disable-line
    loadingText, // eslint-disable-line
    errorLargeFile,
    errorUnsupportedFile,
    errorAmountLimit, // eslint-disable-line
    deleteButton, // eslint-disable-line
    fileListAriaLabel, // eslint-disable-line
    ...props
  } = extendedProps

  const spacingClasses = createSpacingClasses(props)

  const { files, setFiles, setInternalFiles, getExistingFile } =
    useUpload(id)

  const filesRef = React.useRef<UploadFile[]>(files)
  React.useEffect(() => {
    filesRef.current = files
  }) // keep our ref updated on every re-render

  return (
    <UploadContext.Provider
      value={{
        ...extendedProps,
        id,
        onInputUpload,
      }}
    >
      <Provider skeleton={skeleton}>
        <UploadDropzone
          className={classnames('dnb-upload', spacingClasses, className)}
          {...props}
        >
          <UploadInfo />

          <UploadFileInput />

          <UploadFileList />
        </UploadDropzone>
      </Provider>
    </UploadContext.Provider>
  )

  function onInputUpload(newFiles: UploadFile[]) {
    const files = filesRef.current
    const mergedFiles = [
      ...files,
      ...newFiles.map((fileItem) => {
        const { file } = fileItem

        const existingFile = getExistingFile(file, files)

        fileItem.exists = Boolean(existingFile)
        fileItem.id = fileItem.exists ? existingFile.id : makeUniqueId()

        return fileItem
      }),
    ]

    const verifiedFiles = verifyFiles(
      mergedFiles.filter(({ exists }) => !exists),
      {
        fileMaxSize,
        acceptedFileTypes,
        errorUnsupportedFile,
        errorLargeFile,
      }
    )

    const validFiles = [...verifiedFiles].slice(0, filesAmountLimit)

    setFiles(validFiles)
    setInternalFiles(mergedFiles)

    if (typeof onChange === 'function') {
      onChange({ files: validFiles })
    }

    return validFiles
  }
}

Upload.useUpload = useUpload

Upload._supportsSpacingProps = true

export default Upload
