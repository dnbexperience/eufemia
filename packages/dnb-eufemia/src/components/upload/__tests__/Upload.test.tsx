import React, { useEffect } from 'react'
import {
  fireEvent,
  render,
  screen,
  waitFor,
  renderHook,
} from '@testing-library/react'
import Upload from '../Upload'
import nbNO from '../../../shared/locales/nb-NO'
import enGB from '../../../shared/locales/en-GB'
import { createMockFile } from './testHelpers'
import { loadScss, axeComponent } from '../../../core/jest/jestSetup'
import { UploadAllProps } from '../types'
import useUpload from '../useUpload'
import Provider from '../../../shared/Provider'

const nb = nbNO['nb-NO'].Upload
const en = enGB['en-GB'].Upload

global.URL.createObjectURL = jest.fn(() => 'url')

const defaultProps: UploadAllProps = {
  id: 'id',
  acceptedFileTypes: ['png'],
}

describe('Upload', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders the component', () => {
    render(<Upload {...defaultProps} />)

    expect(document.querySelector('.dnb-upload')).toBeInTheDocument()
  })

  it('renders the upload file input section', () => {
    render(<Upload {...defaultProps} />)

    expect(
      document.querySelector('.dnb-upload__file-input')
    ).toBeInTheDocument()
  })

  describe('Text', () => {
    it('renders the title', () => {
      render(<Upload {...defaultProps} />)

      expect(screen.queryByText(nb.title)).toBeInTheDocument()
    })

    it('renders the custom title', () => {
      const customTitle = 'custom title'

      render(<Upload {...defaultProps} title={customTitle} />)

      expect(screen.queryByText(customTitle)).toBeInTheDocument()
    })

    it('renders the text', () => {
      render(<Upload {...defaultProps} />)

      expect(screen.queryByText(nb.text)).toBeInTheDocument()
    })

    it('renders the custom text', () => {
      const customText = 'custom text'

      render(<Upload {...defaultProps} text={customText} />)

      expect(screen.queryByText(customText)).toBeInTheDocument()
    })

    it('renders the format description', () => {
      render(<Upload {...defaultProps} />)

      expect(
        screen.queryByText(nb.fileTypeDescription)
      ).toBeInTheDocument()
    })

    it('renders the custom format description', () => {
      const customFormatDescription = 'custom formats description'

      render(
        <Upload
          {...defaultProps}
          fileTypeDescription={customFormatDescription}
        />
      )

      expect(
        screen.queryByText(customFormatDescription)
      ).toBeInTheDocument()
    })

    it('does not render text when text is false', () => {
      render(<Upload {...defaultProps} text={false} />)

      expect(screen.queryByText(nb.text)).not.toBeInTheDocument()
    })

    it('does not render text when text is empty string', () => {
      render(<Upload {...defaultProps} text="" />)

      expect(screen.queryByText(nb.text)).not.toBeInTheDocument()
    })

    it('does not render title when title is false', () => {
      render(<Upload {...defaultProps} title={false} />)

      expect(screen.queryByText(nb.title)).not.toBeInTheDocument()
    })

    it('does not render title when title is empty string', () => {
      render(<Upload {...defaultProps} title="" />)

      expect(screen.queryByText(nb.title)).not.toBeInTheDocument()
    })

    it('does not render fileTypeDescription when acceptedFileTypes is empty', () => {
      const acceptedFileTypes = []

      render(
        <Upload {...defaultProps} acceptedFileTypes={acceptedFileTypes} />
      )

      expect(
        screen.queryByText(nb.fileTypeDescription)
      ).not.toBeInTheDocument()
    })

    it('does not render the accepted file types table when acceptedFileTypes is an array of strings', () => {
      const acceptedFileTypes = ['jpg', 'png']

      render(
        <Upload {...defaultProps} acceptedFileTypes={acceptedFileTypes} />
      )

      expect(
        screen.queryByText(nb.fileTypeTableCaption)
      ).not.toBeInTheDocument()
    })

    it('renders the accepted file types table when acceptedFileTypes is an array of objects', () => {
      const acceptedFileTypes = [
        { fileType: 'jpg', fileMaxSize: 1 },
        { fileType: 'png', fileMaxSize: 2 },
      ]

      render(
        <Upload {...defaultProps} acceptedFileTypes={acceptedFileTypes} />
      )

      expect(
        screen.queryByText(nb.fileTypeTableCaption)
      ).toBeInTheDocument()
    })

    it('renders the correct grouping of file types based on file max size', () => {
      const fileMaxSize = 99

      const fileMaxSize1Types = [
        {
          fileType: 'jpg',
          fileMaxSize: 1,
        },
        {
          fileType: 'doc',
          fileMaxSize: 1,
        },
        {
          fileType: 'svg',
          fileMaxSize: 1,
        },
        {
          fileType: 'gif',
          fileMaxSize: 1,
        },
      ]

      const fileMaxSize2Types = [
        {
          fileType: 'docx',
          fileMaxSize: 2,
        },
      ]

      const fileMaxSizeUndefinedTypes = [
        {
          fileType: 'odt',
        },
        {
          fileType: 'pdf',
        },
      ]

      const fileMaxSizeDisabledTypes = [
        {
          fileType: 'text',
          fileMaxSize: false,
        },
        {
          fileType: 'txt',
          fileMaxSize: 0,
        },
      ]

      const fileMaxSizeSameTypes = [
        {
          fileType: 'zip',
          fileMaxSize,
        },
      ]

      const acceptedFileTypes = [
        ...fileMaxSize1Types,
        ...fileMaxSize2Types,
        ...fileMaxSizeUndefinedTypes,
        ...fileMaxSizeDisabledTypes,
        ...fileMaxSizeSameTypes,
      ]

      render(
        <Upload
          {...defaultProps}
          fileMaxSize={fileMaxSize}
          acceptedFileTypes={acceptedFileTypes}
        />
      )

      expect(screen.queryByText('ODT, PDF, ZIP')).toBeInTheDocument()
      expect(screen.queryByText('DOCX')).toBeInTheDocument()
      expect(screen.queryByText('DOC, GIF, JPG, SVG')).toBeInTheDocument()
      expect(screen.queryByText('TEXT, TXT')).toBeInTheDocument()
    })

    it('renders the custom accepted format', () => {
      const acceptedFileTypes = ['png', 'jpg']

      render(
        <Upload {...defaultProps} acceptedFileTypes={acceptedFileTypes} />
      )

      const formattedFileTypes = acceptedFileTypes.join(', ').toUpperCase()

      expect(screen.queryByText(formattedFileTypes)).toBeInTheDocument()
    })

    it('renders the file size description', () => {
      render(<Upload {...defaultProps} />)

      expect(
        screen.queryByText(nb.fileSizeDescription)
      ).toBeInTheDocument()
    })

    it('renders the custom file size description', () => {
      const fileSizeDescription = 'file size description'

      render(
        <Upload
          {...defaultProps}
          fileSizeDescription={fileSizeDescription}
        />
      )

      expect(screen.queryByText(fileSizeDescription)).toBeInTheDocument()
    })

    it('renders the file size', () => {
      const fileMaxSize = 2
      render(<Upload {...defaultProps} fileMaxSize={fileMaxSize} />)

      expect(
        screen.queryByText(
          String(nb.fileSizeContent).replace(
            '%size',
            fileMaxSize.toString()
          )
        )
      ).toBeInTheDocument()
    })

    it('renders the custom file size', () => {
      const fileMaxSize = 2
      const fileSizeContent = '%size custom'

      render(
        <Upload
          {...defaultProps}
          fileMaxSize={fileMaxSize}
          fileSizeContent={fileSizeContent}
        />
      )

      expect(
        screen.queryByText(
          String(fileSizeContent).replace('%size', `${fileMaxSize}`)
        )
      ).toBeInTheDocument()
    })

    it('renders no file size when fileMaxSize is disabled', () => {
      const fileMaxSize = 0
      render(<Upload {...defaultProps} fileMaxSize={fileMaxSize} />)

      expect(
        screen.queryByText(
          String(nb.fileSizeContent).replace(
            '%size',
            fileMaxSize.toString()
          )
        )
      ).not.toBeInTheDocument()
    })

    it('renders no custom file size when fileMaxSize is disabled', () => {
      const fileMaxSize = false
      const fileSizeContent = '%size custom'

      render(
        <Upload
          {...defaultProps}
          fileMaxSize={fileMaxSize}
          fileSizeContent={fileSizeContent}
        />
      )

      expect(
        screen.queryByText(
          String(fileSizeContent).replace('%size', `${fileMaxSize}`)
        )
      ).not.toBeInTheDocument()
    })

    it('renders the file amount limit description', () => {
      const filesAmountLimit = 2
      render(
        <Upload {...defaultProps} filesAmountLimit={filesAmountLimit} />
      )

      expect(
        screen.queryByText(nb.fileAmountDescription)
      ).toBeInTheDocument()
      expect(
        screen.queryByText(String(filesAmountLimit))
      ).toBeInTheDocument()
    })

    it('renders the upload file input section button text', () => {
      render(<Upload {...defaultProps} />)

      expect(screen.queryByText(nb.buttonText)).toBeInTheDocument()
    })

    it('renders the upload file input section button custom text', () => {
      const buttonText = 'upload button text'

      render(<Upload {...defaultProps} buttonText={buttonText} />)

      expect(screen.queryByText(buttonText)).toBeInTheDocument()
    })

    it('should support locale prop', () => {
      const { rerender } = render(<Upload {...defaultProps} />)

      expect(screen.queryByText(nb.title)).toBeInTheDocument()

      rerender(<Upload {...defaultProps} locale="en-GB" />)

      expect(screen.queryByText(en.title)).toBeInTheDocument()
    })

    it('should support locale from provider', () => {
      const { rerender } = render(
        <Provider>
          <Upload {...defaultProps} />
        </Provider>
      )

      expect(screen.queryByText(nb.title)).toBeInTheDocument()

      rerender(
        <Provider locale="en-GB">
          <Upload {...defaultProps} />
        </Provider>
      )

      expect(screen.queryByText(en.title)).toBeInTheDocument()

      rerender(
        <Provider locale="nb-NO">
          <Upload {...defaultProps} />
        </Provider>
      )

      expect(screen.queryByText(nb.title)).toBeInTheDocument()
    })

    it('should support spacing props', () => {
      render(<Upload top="2rem" {...defaultProps} />)

      const element = document.querySelector('.dnb-upload')
      const attributes = Array.from(element.attributes).map(
        (attr) => attr.name
      )

      expect(attributes).toEqual(['class'])
      expect(Array.from(element.classList)).toEqual(
        expect.arrayContaining(['dnb-space', 'dnb-space__top--large'])
      )
    })

    it('renders the singular upload button text when filesAmountLimit is 1', () => {
      render(<Upload {...defaultProps} filesAmountLimit={1} />)

      expect(screen.queryByText(nb.buttonText)).not.toBeInTheDocument()
      expect(screen.queryByText(nb.buttonTextSingular)).toBeInTheDocument()
    })

    it('renders the singular text when filesAmountLimit is 1', () => {
      render(<Upload {...defaultProps} filesAmountLimit={1} />)

      expect(screen.queryByText(nb.text)).not.toBeInTheDocument()
      expect(screen.queryByText(nb.textSingular)).toBeInTheDocument()
    })
  })

  it('will only accept one file if filesAmountLimit is 1', async () => {
    const id = 'filesAmountLimit'

    const { result } = renderHook(useUpload, { initialProps: id })

    render(<Upload {...defaultProps} id={id} filesAmountLimit={1} />)

    const getRootElement = () => document.querySelector('.dnb-upload')

    const element = getRootElement()
    const file1 = createMockFile('fileName-1.png', 100, 'image/png')
    const file2 = createMockFile('fileName-2.png', 100, 'image/png')

    await waitFor(() =>
      fireEvent.drop(element, {
        dataTransfer: { files: [file1, file2] },
      })
    )

    await waitFor(() =>
      fireEvent.drop(element, {
        dataTransfer: { files: [file2, file2] },
      })
    )

    const [firstItem] = Array.from(element.querySelectorAll('li'))
    const deleteButton = firstItem.querySelector('button')

    expect(result.current.files.length).toBe(1)
    expect(result.current.files).toEqual([
      { file: file1, id: expect.any(String), exists: false },
    ])
    expect(
      screen.queryByText(nb.errorAmountLimit.replace('%amount', '1'))
    ).toBeInTheDocument()
    expect(result.current.internalFiles.length).toBe(3)

    fireEvent.click(deleteButton)

    await waitFor(() => {
      expect(
        element.querySelector(
          '.dnb-upload__file-input-area .dnb-form-status'
        )
      ).not.toBeInTheDocument()
    })

    expect(
      screen.queryByRole('button', {
        name: nb.buttonTextSingular,
      })
    ).not.toHaveAttribute('disabled')
  })

  it('will accept same file only once', async () => {
    const id = 'only-once'

    const { result } = renderHook(useUpload, { initialProps: id })

    render(<Upload {...defaultProps} id={id} />)

    const getRootElement = () => document.querySelector('.dnb-upload')

    const element = getRootElement()
    const file1 = createMockFile('fileName-1.png', 100, 'image/png')
    const file2 = createMockFile('fileName-2.png', 100, 'image/png')

    await waitFor(() =>
      fireEvent.drop(element, {
        dataTransfer: { files: [file1] },
      })
    )

    await waitFor(() =>
      fireEvent.drop(element, {
        dataTransfer: { files: [file1, file2] },
      })
    )

    expect(result.current.files.length).toBe(2)
    expect(result.current.files).toEqual([
      { file: file1, id: expect.any(String), exists: false },
      { file: file2, id: expect.any(String), exists: false },
    ])
    expect(result.current.internalFiles.length).toBe(3)
    expect(result.current.internalFiles).toEqual([
      { file: file1, id: expect.any(String), exists: false },
      { file: file1, id: expect.any(String), exists: true },
      { file: file2, id: expect.any(String), exists: false },
    ])
  })

  it('will highlight same file', async () => {
    const id = 'highlight'

    renderHook(useUpload, { initialProps: id })

    render(<Upload {...defaultProps} id={id} />)

    const getRootElement = () => document.querySelector('.dnb-upload')

    const element = getRootElement()
    const file1 = createMockFile('fileName-1.png', 100, 'image/png')
    const file2 = createMockFile('fileName-2.png', 100, 'image/png')

    await waitFor(() =>
      fireEvent.drop(element, {
        dataTransfer: { files: [file1] },
      })
    )

    await waitFor(() =>
      fireEvent.drop(element, {
        dataTransfer: { files: [file1, file2] },
      })
    )

    expect(
      element.querySelectorAll('.dnb-upload__file-cell--highlight')
    ).toHaveLength(1)
    expect(
      Array.from(
        element.querySelectorAll('.dnb-upload__file-cell')[0].classList
      )
    ).toEqual(expect.arrayContaining(['dnb-upload__file-cell--highlight']))
  })

  describe('useUpload', () => {
    it('calls uses the useUpload hook to store files', async () => {
      const validationFunction = jest.fn()

      const file = createMockFile('fileName.png', 100, 'image/png')

      const id = 'random-id'

      const expectedResult = [
        { file, id: expect.any(String), exists: false },
      ]

      render(<Upload {...defaultProps} id={id} />)

      const inputElement = document.querySelector(
        '.dnb-upload__file-input'
      )

      await waitFor(() =>
        fireEvent.change(inputElement, {
          target: { files: [file] },
        })
      )

      const MockComponent = () => {
        const { files } = useUpload(id)
        useEffect(() => validationFunction(files), [])

        return <div />
      }

      render(<MockComponent />)

      expect(validationFunction).toHaveBeenCalledWith(expectedResult)
    })

    it('renders the list of files', async () => {
      const files = [
        { file: createMockFile('fileName.png', 100, 'image/png') },
        { file: createMockFile('fileName2.png', 100, 'image/png') },
        { file: createMockFile('fileName3.png', 100, 'image/png') },
      ]

      const id = 'random-id'

      render(<Upload {...defaultProps} id={id} />)

      const MockComponent = () => {
        const { setFiles } = useUpload(id)

        useEffect(() => setFiles(files), [])

        return <div />
      }

      render(<MockComponent />)

      const fileCells = document.querySelectorAll('.dnb-upload__file-cell')

      expect(fileCells.length).toBe(files.length)
    })

    it('shows no files', async () => {
      const files = []

      const id = 'random-id3'

      render(<Upload {...defaultProps} id={id} />)

      const MockComponent = () => {
        const { setFiles } = useUpload(id)

        useEffect(() => setFiles(files), [])

        return <div />
      }
      render(<MockComponent />)

      const emptyFileCell = document.querySelector(
        '.dnb-upload__file-cell'
      )

      expect(emptyFileCell).not.toBeInTheDocument()
    })

    it('shows the file when the file is added', async () => {
      const files = [
        { file: createMockFile('fileName.png', 100, 'image/png') },
      ]

      const id = 'random-id2'

      render(<Upload {...defaultProps} id={id} />)

      const MockComponent = () => {
        const { setFiles } = useUpload(id)

        useEffect(() => setFiles(files), [])

        return <div />
      }

      const emptyFileCell = document.querySelector(
        '.dnb-upload__file-cell'
      )

      expect(emptyFileCell).not.toBeInTheDocument()

      render(<MockComponent />)

      const fileCell = document.querySelector('.dnb-upload__file-cell')

      expect(fileCell).toBeInTheDocument()
    })

    it('removes the file from the list when clicking delete', () => {
      const files = [
        { file: createMockFile('fileName.png', 100, 'image/png') },
      ]

      const id = 'random-id3'

      render(<Upload {...defaultProps} id={id} />)
      const MockComponent = () => {
        const { setFiles } = useUpload(id)

        useEffect(() => setFiles(files), [])

        return <div />
      }

      render(<MockComponent />)

      const fileCell = document.querySelector('.dnb-upload__file-cell')

      expect(fileCell).toBeInTheDocument()

      const deleteButton = screen.queryByRole('button', {
        name: nb.deleteButton,
      })

      fireEvent.click(deleteButton)

      expect(
        document.querySelector('.dnb-upload__file-cell')
      ).not.toBeInTheDocument()
    })

    it('sets focus on choose button when clicking delete', () => {
      const files = [
        { file: createMockFile('fileName.png', 100, 'image/png') },
      ]

      const id = 'random-id3'

      render(<Upload {...defaultProps} id={id} />)
      const MockComponent = () => {
        const { setFiles } = useUpload(id)

        useEffect(() => setFiles(files), [])

        return <div />
      }

      render(<MockComponent />)

      const deleteButton = screen.queryByRole('button', {
        name: nb.deleteButton,
      })

      fireEvent.click(deleteButton)

      expect(document.activeElement).toBe(
        screen.queryByRole('button', {
          name: nb.buttonText,
        })
      )
    })

    it('returns the file size error message', async () => {
      const file1 = createMockFile('fileName1.png', 100000000, 'image/png')

      const fileMaxSize = 1
      const errorMessage = 'error message %size'

      render(
        <Upload
          {...defaultProps}
          fileMaxSize={fileMaxSize}
          errorLargeFile={errorMessage}
        />
      )

      const inputElement = document.querySelector(
        '.dnb-upload__file-input'
      )

      await waitFor(() =>
        fireEvent.change(inputElement, {
          target: { files: [file1] },
        })
      )

      expect(
        screen.queryByText(`error message ${fileMaxSize}`)
      ).toBeInTheDocument()
    })

    it('returns no file size error message when fileMaxSize is disabled', async () => {
      const file1 = createMockFile('fileName1.png', 100000000, 'image/png')

      const fileMaxSize = 0
      const errorMessage = 'error message %size'

      render(
        <Upload
          {...defaultProps}
          fileMaxSize={fileMaxSize}
          errorLargeFile={errorMessage}
        />
      )

      const inputElement = document.querySelector(
        '.dnb-upload__file-input'
      )

      await waitFor(() =>
        fireEvent.change(inputElement, {
          target: { files: [file1] },
        })
      )

      expect(
        screen.queryByText(`error message ${fileMaxSize}`)
      ).not.toBeInTheDocument()
    })
  })

  describe('events', () => {
    it('will call onChange when file gets added or removed', async () => {
      const id = 'onChange'
      const onChange = jest.fn()

      render(<Upload {...defaultProps} id={id} onChange={onChange} />)

      const inputElement = document.querySelector(
        '.dnb-upload__file-input'
      )
      const file1 = createMockFile('fileName-1.png', 100, 'image/png')

      await waitFor(() =>
        fireEvent.change(inputElement, {
          target: { files: [file1] },
        })
      )

      expect(onChange).toHaveBeenCalledTimes(1)
      expect(onChange).toHaveBeenCalledWith({
        files: [{ file: file1, id: expect.any(String), exists: false }],
      })

      const deleteButton = screen.queryByRole('button', {
        name: nb.deleteButton,
      })

      await waitFor(() => fireEvent.click(deleteButton))

      expect(onChange).toHaveBeenCalledTimes(2)
      expect(onChange).toHaveBeenCalledWith({ files: [] })
    })

    it('will call onFileDelete when file gets removed', async () => {
      const id = 'onFileDelete'
      const onFileDelete = jest.fn()

      render(
        <Upload {...defaultProps} id={id} onFileDelete={onFileDelete} />
      )

      const inputElement = document.querySelector(
        '.dnb-upload__file-input'
      )
      const file1 = createMockFile('fileName-1.png', 100, 'image/png')

      await waitFor(() =>
        fireEvent.change(inputElement, {
          target: { files: [file1] },
        })
      )

      const deleteButton = screen.queryByRole('button', {
        name: nb.deleteButton,
      })

      await waitFor(() => fireEvent.click(deleteButton))

      expect(onFileDelete).toHaveBeenCalledTimes(1)
      expect(onFileDelete).toHaveBeenCalledWith({
        fileItem: { file: file1, id: expect.any(String), exists: false },
      })
    })
  })
})

describe('Upload aria', () => {
  it('should validate', async () => {
    const Component = render(<Upload {...defaultProps} />)
    expect(await axeComponent(Component)).toHaveNoViolations()
  })
})

describe('Upload scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-upload-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
