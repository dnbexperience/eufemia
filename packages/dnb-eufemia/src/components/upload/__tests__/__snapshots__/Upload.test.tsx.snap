// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload scss have to match snapshot 1`] = `
"/*
* DNB Upload
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
.dnb-upload {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-basis);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * Ensure consistency and use the same as HTML reset -> html {...}
  * between base and code package
  */
  -moz-tab-size: 4;
  tab-size: 4;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  word-break: break-word;
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-upload *,
  .dnb-upload ::before,
  .dnb-upload ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-upload ::before,
  .dnb-upload ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
* Upload component
*
*/
.dnb-upload {
  display: flex;
  flex-direction: column;
  padding: var(--spacing-medium);
  background-color: var(--color-white);
  position: relative; }
  .dnb-upload__outline {
    content: '';
    inset: 0;
    stroke: var(--color-sea-green);
    position: absolute;
    height: 100%;
    width: 100%;
    border-radius: 0.25rem; }
  .dnb-upload__condition-list {
    margin-top: var(--spacing-small); }
    .dnb-upload__condition-list__label {
      font-weight: var(--font-weight-medium);
      margin-right: var(--spacing-x-small); }
  .dnb-upload__text.dnb-p {
    color: var(--color-black-55); }
  .dnb-upload__file-input {
    visibility: hidden; }
  .dnb-upload__file-list {
    padding: 0;
    margin-bottom: 0;
    position: relative;
    list-style: none; }
  .dnb-upload__file-list::before, .dnb-upload__file-cell::after {
    content: '';
    position: absolute;
    inset: 0;
    height: 1px;
    background-color: var(--color-black-8); }
  .dnb-upload__file-cell {
    padding: var(--spacing-small) 0;
    position: relative; }
    .dnb-upload__file-cell::after {
      top: auto; }
    .dnb-upload__file-cell__content {
      display: flex;
      flex-direction: row;
      justify-content: space-between;
      align-items: center; }
      .dnb-upload__file-cell__content__left {
        display: flex;
        flex-direction: row;
        align-items: center; }
    .dnb-upload__file-cell--warning .dnb-upload__file-cell__content__left .dnb-icon {
      color: var(--color-fire-red); }
    .dnb-upload__file-cell__text-container {
      display: flex;
      flex-direction: column;
      margin-left: var(--spacing-small); }
      .dnb-upload__file-cell__text-container--loading {
        font-size: var(--font-size-basis); }
    .dnb-upload__file-cell__subtitle.dnb-p {
      color: var(--color-black-55); }
"
`;
